api-server/src/server/boot/challenge.js

```diff
@@ -429,35 +502,178 @@ async function backendChallengeCompleted(req, res, next) {
-  );
+    try {
+      await user.getCompletedChallenges$().toPromise();
+    } catch (e) {
+      return next(e);
+    }
 
-  user.updateAttributes(updateData, err => {
-    if (err) {
-      return next(err);
+    try {
+      const examFromDb = await Exam.findById(challengeId);
+      if (!examFromDb) {
+        res.status(500);
+        throw new Error(
+          `An error occurred trying to get the exam from the database.`
+        );
+      }
+
+      // This is cause there was struggles validating the exam directly from the db/loopback
+      const examJson = JSON.parse(JSON.stringify(examFromDb));
+
+      const validExamFromDbSchema = validateExamFromDbSchema(examJson);
+
+      if (validExamFromDbSchema.error) {
+        res.status(500);
+        log(validExamFromDbSchema.error);
+        throw new Error(
+          `An error occurred validating the exam information from the database.`
+        );
+      }
+
+      const { prerequisites, numberOfQuestionsInExam, title } = examJson;
+
+      // Validate User has completed prerequisite challenges
+      prerequisites?.forEach(prerequisite => {
+        const prerequisiteCompleted = user.completedChallenges.find(
+          challenge => challenge.id === prerequisite.id
+        );
+
+        if (!prerequisiteCompleted) {
+          res.status(403);
+          throw new Error(
+            `You have not completed the required challenges to start the '${title}'.`
+          );
+        }
+      });
+
+      const randomizedExam = generateRandomExam(examJson);
+
+      const validGeneratedExamSchema = validateGeneratedExamSchema(
+        randomizedExam,
+        numberOfQuestionsInExam
+      );
+
+      if (validGeneratedExamSchema.error) {
+        res.status(500);
+        log(validGeneratedExamSchema.error);
+        throw new Error(`An error occurred trying to randomize the exam.`);
+      }
+
+      return res.send({ generatedExam: randomizedExam });
+    } catch (err) {
+      log(err);
+      return res.send({ error: err.message });
     }
+  };
+}
 
-    return res.json({
-      alreadyCompleted,
-      points: alreadyCompleted ? user.points : user.points + 1,
-      completedDate: completedChallenge.completedDate
-    });
-  });
+function createExamChallengeCompleted(app) {
+  const { Exam } = app.models;
+
+  return async function examChallengeCompleted(req, res, next) {
+    const { body = {}, user } = req;
+
+    try {
+      await user.getCompletedChallenges$().toPromise();
+    } catch (e) {
+      return next(e);
+    }
+
+    const { userCompletedExam = [], id } = body;
+
+    try {
+      const examFromDb = await Exam.findById(id);
+      if (!examFromDb) {
+        res.status(500);
+        throw new Error(
+          `An error occurred tryng to get the exam from the database.`
+        );
+      }
+
+      // This is cause there was struggles validating the exam directly from the db/loopback
+      const examJson = JSON.parse(JSON.stringify(examFromDb));
```

Comment:
I was getting this: `error: [Error [ValidationError]: "prerequisites[0].__cachedRelations" is not allowed]` - not quite sure what was happening. The `__cachedRelations` is something that loopback adds to, or when pulling info from, the database I think - or both? I don't know - but trying to validate the info right from the database included that field is seems - even though it isn't shown when I look at the data in the database. When I stringify and parse it, it removes it? Something like that. We should be able to remove this in the new API I think.

Reply:
Right, that kind of makes sense, but I don't see why passing it through JSON.stringify helps.

For example, if I do this:
```
JSON.parse(JSON.stringify([{__thing: 'value'}]))
```
I get back `[{__thing: 'value'}]`.

Reply:
I wonder if this is some nonsense with non-enumerable properties.

Reply:
Oh, yeah, that's probably it.

I'm fine with this approach since we don't seem to be losing anything we care about. If we have time to confirm, that'd be nice, but it's not a huge deal either way.

Reply:
I did some digging.

Loopback appears to depend on `loopback-datasource-juggler` in our pnpm-lock, and that's where I can see the `Object.defineProperty()` call.

https://github.com/loopbackio/loopback-datasource-juggler/blob/f7ded61ad52984cd85c8655601d4c6e1a35b6eba/lib/model.js#L114

Reply:
And there it is. Nice catch, @naomi-lgbt!